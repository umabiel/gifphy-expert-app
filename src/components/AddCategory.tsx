import React, {
  ChangeEvent,
  Dispatch,
  FormEvent,
  Fragment,
  SetStateAction,
  useState,
} from "react";

import PropTypes from "prop-types";

interface IProps {
  setCategories: Dispatch<SetStateAction<string[]>>;
}

export const AddCategory = ({ setCategories }: IProps) => {
  const [inputValue, setInputValue] = useState("");
  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (inputValue.trim().length > 2) {
      setCategories((animes) => [inputValue, ...animes]);
      setInputValue("");
    }
  };
  return (
    <Fragment>
      <form onSubmit={handleSubmit}>
        <input type="text" value={inputValue} onChange={handleInputChange} />
      </form>
    </Fragment>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
