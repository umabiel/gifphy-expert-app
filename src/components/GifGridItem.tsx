import { IGifs } from "../core/oop/interfaces/IGifs";

export const GifGridItem = ({ id, title, url }: IGifs) => {
  return (
    <div className="card">
      <p>{title}</p>
      <img src={url} alt={title} />
    </div>
  );
};
