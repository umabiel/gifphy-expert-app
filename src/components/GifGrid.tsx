import React, { Fragment } from "react";
import { IGifs } from "../core/oop/interfaces/IGifs";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from "prop-types";

interface IProps {
  category: string;
}

export const GifGrid = ({ category = "One Punch" }: IProps) => {
  // const initState: IGifs[] = [];
  // const [images, setImages] = useState(initState);
  const { data: images, loading } = useFetchGifs(category);

  return (
    <Fragment>
      <h3>{category}</h3>
      {loading && <p>Loading...</p>}
      <div className="card-grid">
        {images.map((x: IGifs) => (
          <GifGridItem key={x.id} {...x} />
        ))}
      </div>
    </Fragment>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
