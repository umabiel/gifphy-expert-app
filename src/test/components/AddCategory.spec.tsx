import { shallow, ShallowWrapper } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe("Pruebas en AddCategory", () => {
  const setCategories = jest.fn();
  let wrapper: ShallowWrapper; // = shallow(<AddCategory setCategories={setCategories} />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });
  test("Debe renderizar correctamente <AddCategory />", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe cambiar en contenido de la caja de texto", () => {
    const input = wrapper.find("input");
    const value = "Hola Mundo!";

    input.simulate("change", { target: { value } });
    // expect(input.text().trim).toBe(value);
  });

  test("No se debe pedir informacion si el input esta vacio", () => {
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
    expect(setCategories).not.toHaveBeenCalled();
  });

  test("Debe de llamar a la funcion setCategories y limpiar la caja de texto", () => {
    const value = "Hola Mundo!";
    // 1. Simular entrada en input
    wrapper.find("input").simulate("change", { target: { value } });
    // 2. Simular envio de formulario
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
    // 3. Llamar setCategories de manera correcta
    expect(setCategories).toHaveBeenCalledTimes(1); // Cantidad de veces que se debio llamar
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
    // 4. Limpiar la caja de texto
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
