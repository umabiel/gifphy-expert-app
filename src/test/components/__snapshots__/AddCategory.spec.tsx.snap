// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AddCategory Debe renderizar correctamente <AddCategory /> 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
</Fragment>
`;
